File: js/src/v2/session/keystore.ts
/**
 * AUN E2EE V2: 设备密钥存储（IndexedDB 持久化）
 *
 * 与 Python `aun_core.v2.keystore.V2KeyStore` 行为对齐：
 * - 每个 `(device_id)` 持有一对 IK 与若干 SPK（按 spk_id 索引）
 * - SPK 按 created_at 排序，支持 `loadCurrentSPK` / `listRecentSPKIds(n)`
 *
 * 浏览器目标：所有 IO 是 async（IndexedDB 事务）。
 */

export const V2_DB_NAME = 'aun_v2';
export const V2_DB_VERSION = 3;
export const V2_STORE_NAME = 'v2_device_keys';
export const V2_INDEX_BY_DEVICE_TYPE_CREATED = 'by_device_type_created';

type KeyType = 'ik' | 'spk' | 'group_spk' | 'group_identity' | 'spk_uploaded' | 'group_spk_uploaded';

interface V2KeyRecord {
  device_id: string;
  key_type: KeyType;
  group_id: string;
  key_id: string; // IK 主记录用空串，IK fallback alias 用 sha256:<16hex>
  private_key: Uint8Array;
  public_key: Uint8Array;
  created_at: number; // ms
}

async function spkIdForPubDer(pubDer: Uint8Array): Promise<string> {
  const buf = await crypto.subtle.digest('SHA-256', pubDer.slice().buffer);
  const arr = new Uint8Array(buf);
  let hex = '';
  for (let i = 0; i < arr.length; i++) hex += arr[i].toString(16).padStart(2, '0');
  return `sha256:${hex.slice(0, 16)}`;
}

function migrateRecord(raw: Partial<V2KeyRecord>): V2KeyRecord {
  let groupId = String(raw.group_id ?? '');
  let keyId = String(raw.key_id ?? '');
  if ((raw.key_type === 'group_spk' || raw.key_type === 'group_spk_uploaded') && !groupId && keyId.includes('\0')) {
    const parts = keyId.split('\0');
    groupId = parts[0] ?? '';
    keyId = parts.slice(1).join('\0');
  }
  return {
    device_id: String(raw.device_id ?? ''),
    key_type: raw.key_type as KeyType,
    group_id: groupId,
    key_id: keyId,
    private_key: new Uint8Array(raw.private_key ?? new Uint8Array()),
    public_key: new Uint8Array(raw.public_key ?? new Uint8Array()),
    created_at: Number(raw.created_at ?? Date.now()),
  };
}

function createV2Store(db: IDBDatabase): IDBObjectStore {
  const store = db.createObjectStore(V2_STORE_NAME, {
    keyPath: ['device_id', 'key_type', 'group_id', 'key_id'],
  });
  store.createIndex(
    V2_INDEX_BY_DEVICE_TYPE_CREATED,
    ['device_id', 'key_type', 'group_id', 'created_at'],
  );
  return store;
}

function recreateScopeCreatedIndex(store: IDBObjectStore): void {
  if (store.indexNames.contains(V2_INDEX_BY_DEVICE_TYPE_CREATED)) {
    store.deleteIndex(V2_INDEX_BY_DEVICE_TYPE_CREATED);
  }
  store.createIndex(
    V2_INDEX_BY_DEVICE_TYPE_CREATED,
    ['device_id', 'key_type', 'group_id', 'created_at'],
  );
}

/**
 * V2 设备密钥持久化存储。复合主键 [device_id, key_type, group_id, key_id]。
 *
 * 使用 IndexedDB；在浏览器内置 indexedDB 不可用时（如 jsdom）请提前安装
 * `fake-indexeddb/auto` 作为 polyfill（见 `tests/setup.ts`）。
 */
export class V2KeyStore {
  constructor(private readonly db: IDBDatabase) {}

  /** 打开/创建 V2 keystore 数据库。 */
  static async open(dbName: string = V2_DB_NAME): Promise<V2KeyStore> {
    return new Promise((resolve, reject) => {
      const req = indexedDB.open(dbName, V2_DB_VERSION);
      req.onupgradeneeded = (event) => {
        const db = req.result;
        if (!db.objectStoreNames.contains(V2_STORE_NAME)) {
          createV2Store(db);
          return;
        }
        if (event.oldVersion < 2) {
          const tx = req.transaction;
          if (!tx) throw new Error('V2KeyStore.open: missing upgrade transaction');
          const oldStore = tx.objectStore(V2_STORE_NAME);
          const getAllReq = oldStore.getAll();
          getAllReq.onsuccess = () => {
            const records = (getAllReq.result as Array<Partial<V2KeyRecord>>).map(migrateRecord);
            db.deleteObjectStore(V2_STORE_NAME);
            const store = createV2Store(db);
            for (const record of records) store.put(record);
          };
          getAllReq.onerror = () => reject(getAllReq.error);
        } else if (event.oldVersion < 3) {
          const tx = req.transaction;
          if (!tx) throw new Error('V2KeyStore.open: missing upgrade transaction');
          recreateScopeCreatedIndex(tx.objectStore(V2_STORE_NAME));
        }
      };
      req.onsuccess = () => resolve(new V2KeyStore(req.result));
      req.onerror = () => reject(req.error);
      req.onblocked = () => reject(new Error('V2KeyStore.open: blocked'));
    });
  }

  /** 关闭数据库连接（测试或释放资源时使用）。 */
  close(): void {
    this.db.close();
  }

  private store(mode: IDBTransactionMode): IDBObjectStore {
    return this.db.transaction(V2_STORE_NAME, mode).objectStore(V2_STORE_NAME);
  }


  private async _listRecordsByTypeNewestFirst(
    deviceId: string,
    keyType: KeyType,
    groupId: string,
  ): Promise<V2KeyRecord[]> {
    return new Promise((resolve, reject) => {
      const idx = this.store('readonly').index(V2_INDEX_BY_DEVICE_TYPE_CREATED);
      const range = IDBKeyRange.bound(
        [deviceId, keyType, groupId, -Infinity],
        [deviceId, keyType, groupId, Infinity],
      );
      const req = idx.openCursor(range, 'prev');
      const out: V2KeyRecord[] = [];
      req.onsuccess = () => {
        const cursor = req.result;
        if (cursor) {
          out.push(cursor.value as V2KeyRecord);
          cursor.continue();
        } else {
          resolve(out);
        }
      };
      req.onerror = () => reject(req.error);
    });
  }

  // ---------- SPK ----------

  async saveSPK(
    deviceId: string,
    spkId: string,
    priv: Uint8Array,
    pubDer: Uint8Array,
  ): Promise<void> {
    const record: V2KeyRecord = {
      device_id: deviceId,
      key_type: 'spk',
      group_id: '',
      key_id: spkId,
      private_key: priv,
      public_key: pubDer,
      created_at: Date.now(),
    };
    return new Promise((resolve, reject) => {
      const req = this.store('readwrite').put(record);
      req.onsuccess = () => resolve();
      req.onerror = () => reject(req.error);
    });
  }

  async loadSPK(deviceId: string, spkId: string): Promise<Uint8Array | null> {
    return new Promise((resolve, reject) => {
      const req = this.store('readonly').get([deviceId, 'spk', '', spkId]);
      req.onsuccess = () => {
        const r = req.result as V2KeyRecord | undefined;
        // 用 new Uint8Array(...) 拷贝，规避 fake-indexeddb / 跨 realm 实例对象
        resolve(r ? new Uint8Array(r.private_key) : null);
      };
      req.onerror = () => reject(req.error);
    });
  }

  /** 取最新 SPK（按 created_at DESC LIMIT 1）。 */
  async loadCurrentSPK(
    deviceId: string,
  ): Promise<{ spkId: string; priv: Uint8Array; pubDer: Uint8Array } | null> {
    return new Promise((resolve, reject) => {
      const idx = this.store('readonly').index(V2_INDEX_BY_DEVICE_TYPE_CREATED);
      const range = IDBKeyRange.bound(
        [deviceId, 'spk', '', -Infinity],
        [deviceId, 'spk', '', Infinity],
      );
      const req = idx.openCursor(range, 'prev');
      req.onsuccess = () => {
        const cursor = req.result;
        if (!cursor) {
          resolve(null);
          return;
        }
        const r = cursor.value as V2KeyRecord;
        resolve({
          spkId: r.key_id,
          priv: new Uint8Array(r.private_key),
          pubDer: new Uint8Array(r.public_key),
        });
      };
      req.onerror = () => reject(req.error);
    });
  }

  async deleteSPK(deviceId: string, spkId: string): Promise<void> {
    await new Promise<void>((resolve, reject) => {
      const req = this.store('readwrite').delete([deviceId, 'spk', '', spkId]);
      req.onsuccess = () => resolve();
      req.onerror = () => reject(req.error);
    });
    await new Promise<void>((resolve, reject) => {
      const req = this.store('readwrite').delete([deviceId, 'spk_uploaded', '', spkId]);
      req.onsuccess = () => resolve();
      req.onerror = () => reject(req.error);
    });
  }

  async markSPKUploaded(deviceId: string, spkId: string): Promise<void> {
    const record: V2KeyRecord = {
      device_id: deviceId,
      key_type: 'spk_uploaded',
      group_id: '',
      key_id: spkId,
      private_key: new Uint8Array(),
      public_key: new Uint8Array(),
      created_at: Date.now(),
    };
    return new Promise((resolve, reject) => {
      const req = this.store('readwrite').put(record);
      req.onsuccess = () => resolve();
      req.onerror = () => reject(req.error);
    });
  }

  async loadLatestUploadedSPKId(deviceId: string): Promise<string | null> {
    const records = await this._listRecordsByTypeNewestFirst(deviceId, 'spk_uploaded', '');
    for (const record of records) {
      const spkId = record.key_id;
      if (await this.loadSPK(deviceId, spkId)) return spkId;
    }
    return null;
  }

  /** 返回最近 N 代 SPK 的 spk_id（按 created_at DESC）。 */
  async listRecentSPKIds(deviceId: string, n: number): Promise<string[]> {
    if (n <= 0) return [];
    return new Promise((resolve, reject) => {
      const idx = this.store('readonly').index(V2_INDEX_BY_DEVICE_TYPE_CREATED);
      const range = IDBKeyRange.bound(
        [deviceId, 'spk', '', -Infinity],
        [deviceId, 'spk', '', Infinity],
      );
      const req = idx.openCursor(range, 'prev');
      const out: string[] = [];
      req.onsuccess = () => {
        const cursor = req.result;
        if (cursor && out.length < n) {
          out.push((cursor.value as V2KeyRecord).key_id);
          cursor.continue();
        } else {
          resolve(out);
        }
      };
      req.onerror = () => reject(req.error);
    });
  }

  async listExpiredSPKIds(deviceId: string, maxAgeMs: number): Promise<string[]> {
    const cutoff = Date.now() - maxAgeMs;
    return new Promise((resolve, reject) => {
      const idx = this.store('readonly').index(V2_INDEX_BY_DEVICE_TYPE_CREATED);
      const range = IDBKeyRange.bound(
        [deviceId, 'spk', '', -Infinity],
        [deviceId, 'spk', '', cutoff],
        false, true,
      );
      const req = idx.openCursor(range);
      const out: string[] = [];
      req.onsuccess = () => {
        const cursor = req.result;
        if (cursor) {
          out.push((cursor.value as V2KeyRecord).key_id);
          cursor.continue();
        } else {
          resolve(out);
        }
      };
      req.onerror = () => reject(req.error);
    });
  }

  // ---------- Group SPK ----------

  async saveGroupSPK(
    deviceId: string,
    groupId: string,
    spkId: string,
    priv: Uint8Array,
    pubDer: Uint8Array,
  ): Promise<void> {
    const record: V2KeyRecord = {
      device_id: deviceId,
      key_type: 'group_spk',
      group_id: groupId,
      key_id: spkId,
      private_key: priv,
      public_key: pubDer,
      created_at: Date.now(),
    };
    return new Promise((resolve, reject) => {
      const req = this.store('readwrite').put(record);
      req.onsuccess = () => resolve();
      req.onerror = () => reject(req.error);
    });
  }

  async loadGroupSPK(
    deviceId: string,
    groupId: string,
    spkId: string,
  ): Promise<Uint8Array | null> {
    return new Promise((resolve, reject) => {
      const req = this.store('readonly').get([deviceId, 'group_spk', groupId, spkId]);
      req.onsuccess = () => {
        const r = req.result as V2KeyRecord | undefined;
        resolve(r ? new Uint8Array(r.private_key) : null);
      };
      req.onerror = () => reject(req.error);
    });
  }

  /** 取指定群最新 group SPK（按 created_at DESC）。 */
  async loadCurrentGroupSPK(
    deviceId: string,
    groupId: string,
  ): Promise<{ spkId: string; priv: Uint8Array; pubDer: Uint8Array } | null> {
    return new Promise((resolve, reject) => {
      const idx = this.store('readonly').index(V2_INDEX_BY_DEVICE_TYPE_CREATED);
      const range = IDBKeyRange.bound(
        [deviceId, 'group_spk', groupId, -Infinity],
        [deviceId, 'group_spk', groupId, Infinity],
      );
      const req = idx.openCursor(range, 'prev');
      req.onsuccess = () => {
        const cursor = req.result;
        if (!cursor) {
          resolve(null);
          return;
        }
        const r = cursor.value as V2KeyRecord;
        resolve({
          spkId: r.key_id,
          priv: new Uint8Array(r.private_key),
          pubDer: new Uint8Array(r.public_key),
        });
      };
      req.onerror = () => reject(req.error);
    });
  }

  async saveGroupIdentity(
    deviceId: string,
    groupAid: string,
    privateKeyPem: string,
    pubDer: Uint8Array,
  ): Promise<void> {
    const record: V2KeyRecord = {
      device_id: deviceId,
      key_type: 'group_identity',
      group_id: groupAid,
      key_id: '',
      private_key: new TextEncoder().encode(privateKeyPem),
      public_key: pubDer,
      created_at: Date.now(),
    };
    return new Promise((resolve, reject) => {
      const req = this.store('readwrite').put(record);
      req.onsuccess = () => resolve();
      req.onerror = () => reject(req.error);
    });
  }

  async loadGroupIdentity(
    deviceId: string,
    groupAid: string,
  ): Promise<{ privateKeyPem: string; pubDer: Uint8Array } | null> {
    return new Promise((resolve, reject) => {
      const req = this.store('readonly').get([deviceId, 'group_identity', groupAid, '']);
      req.onsuccess = () => {
        const r = req.result as V2KeyRecord | undefined;
        resolve(r ? {
          privateKeyPem: new TextDecoder().decode(r.private_key),
          pubDer: new Uint8Array(r.public_key),
        } : null);
      };
      req.onerror = () => reject(req.error);
    });
  }

  async markGroupSPKUploaded(deviceId: string, groupId: string, spkId: string): Promise<void> {
    const record: V2KeyRecord = {
      device_id: deviceId,
      key_type: 'group_spk_uploaded',
      group_id: groupId,
      key_id: spkId,
      private_key: new Uint8Array(),
      public_key: new Uint8Array(),
      created_at: Date.now(),
    };
    return new Promise((resolve, reject) => {
      const req = this.store('readwrite').put(record);
      req.onsuccess = () => resolve();
      req.onerror = () => reject(req.error);
    });
  }

  async loadLatestUploadedGroupSPKId(deviceId: string, groupId: string): Promise<string | null> {
    const records = await this._listRecordsByTypeNewestFirst(deviceId, 'group_spk_uploaded', groupId);
    for (const record of records) {
      const spkId = record.key_id;
      if (await this.loadGroupSPK(deviceId, groupId, spkId)) return spkId;
    }
    return null;
  }
  // ---------- IK ----------

  async saveIK(deviceId: string, priv: Uint8Array, pubDer: Uint8Array): Promise<void> {
    const record: V2KeyRecord = {
      device_id: deviceId,
      key_type: 'ik',
      group_id: '',
      key_id: '',
      private_key: priv,
      public_key: pubDer,
      created_at: Date.now(),
    };
    const aliasKeyId = await spkIdForPubDer(pubDer);
    const alias: V2KeyRecord = {
      ...record,
      key_id: aliasKeyId,
    };
    return new Promise((resolve, reject) => {
      const tx = this.db.transaction(V2_STORE_NAME, 'readwrite');
      const store = tx.objectStore(V2_STORE_NAME);
      store.put(record);
      store.put(alias);
      tx.oncomplete = () => resolve();
      tx.onerror = () => reject(tx.error);
    });
  }

  async loadIK(
    deviceId: string,
  ): Promise<{ priv: Uint8Array; pubDer: Uint8Array } | null> {
    return new Promise((resolve, reject) => {
      const req = this.store('readonly').get([deviceId, 'ik', '', '']);
      req.onsuccess = () => {
        const r = req.result as V2KeyRecord | undefined;
        resolve(
          r ? { priv: new Uint8Array(r.private_key), pubDer: new Uint8Array(r.public_key) } : null,
        );
      };
      req.onerror = () => reject(req.error);
    });
  }

  async loadIKSPK(
    deviceId: string,
    spkId: string,
  ): Promise<{ priv: Uint8Array; pubDer: Uint8Array } | null> {
    return new Promise((resolve, reject) => {
      const req = this.store('readonly').get([deviceId, 'ik', '', spkId]);
      req.onsuccess = () => {
        const r = req.result as V2KeyRecord | undefined;
        resolve(
          r ? { priv: new Uint8Array(r.private_key), pubDer: new Uint8Array(r.public_key) } : null,
        );
      };
      req.onerror = () => reject(req.error);
    });
  }

  // ---------- 测试用 ----------

  /** 测试用：清空 store。 */
  async _clear(): Promise<void> {
    return new Promise((resolve, reject) => {
      const req = this.store('readwrite').clear();
      req.onsuccess = () => resolve();
      req.onerror = () => reject(req.error);
    });
  }
}

File: python/src/aun_core/v2/keystore.py
"""V2 E2EE 设备密钥存储。

每个 (aid, device_id) 持有一对 IK 和若干 SPK（按 spk_id 索引）。
底层复用 AID 的 aun.db（v2_device_keys 表，DDL 在 sqlite_db.py 中定义）。
"""
from __future__ import annotations

import hashlib
import time
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..keystore.sqlite_db import AIDDatabase


class V2KeyStore:
    """V2 E2EE 设备密钥持久化，复用 AIDDatabase 连接。

    key_type: "ik"、"spk" 或 "group_spk"
    key_id: IK 主记录为空字符串；IK fallback alias 使用 sha256:<16hex>，但仍保持 key_type='ik'
    """

    def __init__(self, db: "AIDDatabase"):
        self._db = db
        self._ensure_schema()

    @staticmethod
    def _ik_spk_id(ik_pub_der: bytes) -> str:
        return "sha256:" + hashlib.sha256(ik_pub_der).hexdigest()[:16]

    def _ensure_schema(self) -> None:
        conn = self._db._get_conn()
        conn.execute(
            """CREATE TABLE IF NOT EXISTS v2_device_keys (
                device_id TEXT NOT NULL,
                key_type TEXT NOT NULL,
                group_id TEXT NOT NULL DEFAULT '',
                key_id TEXT NOT NULL DEFAULT '',
                private_key BLOB NOT NULL,
                public_key BLOB NOT NULL,
                created_at INTEGER NOT NULL,
                PRIMARY KEY (device_id, key_type, group_id, key_id)
            )"""
        )
        rows = conn.execute("PRAGMA table_info(v2_device_keys)").fetchall()
        columns = {str(row[1]) for row in rows}
        pk_columns = [str(row[1]) for row in sorted((r for r in rows if int(r[5] or 0) > 0), key=lambda r: int(r[5]))]
        if "group_id" not in columns or pk_columns != ["device_id", "key_type", "group_id", "key_id"]:
            conn.execute("ALTER TABLE v2_device_keys RENAME TO v2_device_keys_legacy")
            conn.execute(
                """CREATE TABLE v2_device_keys (
                    device_id TEXT NOT NULL,
                    key_type TEXT NOT NULL,
                    group_id TEXT NOT NULL DEFAULT '',
                    key_id TEXT NOT NULL DEFAULT '',
                    private_key BLOB NOT NULL,
                    public_key BLOB NOT NULL,
                    created_at INTEGER NOT NULL,
                    PRIMARY KEY (device_id, key_type, group_id, key_id)
                )"""
            )
            legacy_has_group_id = "group_id" in columns
            select_cols = (
                "device_id, key_type, group_id, key_id, private_key, public_key, created_at"
                if legacy_has_group_id
                else "device_id, key_type, '' AS group_id, key_id, private_key, public_key, created_at"
            )
            legacy_rows = conn.execute(f"SELECT {select_cols} FROM v2_device_keys_legacy").fetchall()
            for device_id, key_type, group_id, key_id, private_key, public_key, created_at in legacy_rows:
                migrated_group_id = str(group_id or "")
                migrated_key_id = str(key_id or "")
                if key_type in ("group_spk", "group_spk_uploaded") and "\0" in migrated_key_id:
                    migrated_group_id, migrated_key_id = migrated_key_id.split("\0", 1)
                if private_key is None or public_key is None:
                    if key_type not in ("spk_uploaded", "group_spk_uploaded"):
                        continue
                    if private_key is None:
                        private_key = b""
                    if public_key is None:
                        public_key = b""
                conn.execute(
                    "INSERT OR REPLACE INTO v2_device_keys "
                    "(device_id, key_type, group_id, key_id, private_key, public_key, created_at) "
                    "VALUES (?, ?, ?, ?, ?, ?, ?)",
                    (device_id, key_type, migrated_group_id, migrated_key_id, private_key, public_key, created_at),
                )
            conn.execute("DROP TABLE v2_device_keys_legacy")
        conn.execute(
            "CREATE INDEX IF NOT EXISTS idx_v2_device_keys_scope_created "
            "ON v2_device_keys(device_id, key_type, group_id, created_at)"
        )
        conn.commit()

    def save_ik(self, device_id: str, ik_priv: bytes, ik_pub_der: bytes) -> None:
        conn = self._db._get_conn()
        now_ms = int(time.time() * 1000)
        conn.execute(
            "INSERT OR REPLACE INTO v2_device_keys (device_id, key_type, group_id, key_id, private_key, public_key, created_at) "
            "VALUES (?, 'ik', '', '', ?, ?, ?)",
            (device_id, ik_priv, ik_pub_der, now_ms),
        )
        conn.execute(
            "INSERT OR REPLACE INTO v2_device_keys (device_id, key_type, group_id, key_id, private_key, public_key, created_at) "
            "VALUES (?, 'ik', '', ?, ?, ?, ?)",
            (device_id, self._ik_spk_id(ik_pub_der), ik_priv, ik_pub_der, now_ms),
        )
        conn.commit()

    def load_ik(self, device_id: str) -> tuple[bytes, bytes] | None:
        conn = self._db._get_conn()
        row = conn.execute(
            "SELECT private_key, public_key FROM v2_device_keys WHERE device_id=? AND key_type='ik' AND group_id='' AND key_id=''",
            (device_id,),
        ).fetchone()
        if row is None:
            return None
        return (bytes(row[0]), bytes(row[1]))

    def load_ik_spk(self, device_id: str, spk_id: str) -> tuple[bytes, bytes] | None:
        conn = self._db._get_conn()
        row = conn.execute(
            "SELECT private_key, public_key FROM v2_device_keys WHERE device_id=? AND key_type='ik' AND group_id='' AND key_id=?",
            (device_id, spk_id),
        ).fetchone()
        if row is None:
            return None
        return (bytes(row[0]), bytes(row[1]))

    def save_spk(self, device_id: str, spk_id: str, spk_priv: bytes, spk_pub_der: bytes) -> None:
        conn = self._db._get_conn()
        conn.execute(
            "INSERT OR REPLACE INTO v2_device_keys (device_id, key_type, group_id, key_id, private_key, public_key, created_at) "
            "VALUES (?, 'spk', '', ?, ?, ?, ?)",
            (device_id, spk_id, spk_priv, spk_pub_der, int(time.time() * 1000)),
        )
        conn.commit()

    def load_spk(self, device_id: str, spk_id: str) -> bytes | None:
        conn = self._db._get_conn()
        row = conn.execute(
            "SELECT private_key FROM v2_device_keys WHERE device_id=? AND key_type='spk' AND group_id='' AND key_id=?",
            (device_id, spk_id),
        ).fetchone()
        if row is None:
            return None
        return bytes(row[0])

    def save_group_spk(self, device_id: str, group_id: str, spk_id: str, spk_priv: bytes, spk_pub_der: bytes) -> None:
        conn = self._db._get_conn()
        conn.execute(
            "INSERT OR REPLACE INTO v2_device_keys (device_id, key_type, group_id, key_id, private_key, public_key, created_at) "
            "VALUES (?, 'group_spk', ?, ?, ?, ?, ?)",
            (device_id, group_id, spk_id, spk_priv, spk_pub_der, int(time.time() * 1000)),
        )
        conn.commit()

    def load_group_spk(self, device_id: str, group_id: str, spk_id: str) -> bytes | None:
        conn = self._db._get_conn()
        row = conn.execute(
            "SELECT private_key FROM v2_device_keys WHERE device_id=? AND key_type='group_spk' AND group_id=? AND key_id=?",
            (device_id, group_id, spk_id),
        ).fetchone()
        if row is None:
            return None
        return bytes(row[0])

    def load_current_group_spk(self, device_id: str, group_id: str) -> tuple[str, bytes, bytes] | None:
        conn = self._db._get_conn()
        row = conn.execute(
            "SELECT key_id, private_key, public_key FROM v2_device_keys "
            "WHERE device_id=? AND key_type='group_spk' AND group_id=? "
            "ORDER BY created_at DESC LIMIT 1",
            (device_id, group_id),
        ).fetchone()
        if row is None:
            return None
        return (str(row[0]), bytes(row[1]), bytes(row[2]))

    def load_current_spk(self, device_id: str) -> tuple[str, bytes, bytes] | None:
        """加载最新的 SPK（按 created_at 降序）。返回 (spk_id, priv, pub) 或 None。"""
        conn = self._db._get_conn()
        row = conn.execute(
            "SELECT key_id, private_key, public_key FROM v2_device_keys "
            "WHERE device_id=? AND key_type='spk' AND group_id='' ORDER BY created_at DESC LIMIT 1",
            (device_id,),
        ).fetchone()
        if row is None:
            return None
        return (row[0], bytes(row[1]), bytes(row[2]))

    def delete_spk(self, device_id: str, spk_id: str) -> None:
        """销毁指定 SPK 的私钥（用于 PFS），并清理对应上传成功标记。"""
        conn = self._db._get_conn()
        conn.execute(
            "DELETE FROM v2_device_keys WHERE device_id=? AND key_type='spk' AND group_id='' AND key_id=?",
            (device_id, spk_id),
        )
        conn.execute(
            "DELETE FROM v2_device_keys WHERE device_id=? AND key_type='spk_uploaded' AND group_id='' AND key_id=?",
            (device_id, spk_id),
        )
        conn.commit()

    def mark_spk_uploaded(self, device_id: str, spk_id: str) -> None:
        """记录指定 P2P SPK 已成功上传到服务端。"""
        conn = self._db._get_conn()
        conn.execute(
            "INSERT OR REPLACE INTO v2_device_keys (device_id, key_type, group_id, key_id, private_key, public_key, created_at) "
            "VALUES (?, 'spk_uploaded', '', ?, ?, ?, ?)",
            (device_id, spk_id, b"", b"", int(time.time() * 1000)),
        )
        conn.commit()

    def load_latest_uploaded_spk_id(self, device_id: str) -> str | None:
        """返回本设备最近一次上传成功且本地私钥仍存在的 P2P SPK ID。"""
        conn = self._db._get_conn()
        row = conn.execute(
            "SELECT marker.key_id FROM v2_device_keys AS marker "
            "WHERE marker.device_id=? AND marker.key_type='spk_uploaded' AND marker.group_id='' "
            "AND EXISTS ("
            "  SELECT 1 FROM v2_device_keys AS key "
            "  WHERE key.device_id=marker.device_id AND key.key_type='spk' AND key.group_id='' AND key.key_id=marker.key_id"
            ") "
            "ORDER BY marker.created_at DESC LIMIT 1",
            (device_id,),
        ).fetchone()
        return str(row[0]) if row is not None else None

    def mark_group_spk_uploaded(self, device_id: str, group_id: str, spk_id: str) -> None:
        """记录指定群 SPK 已成功上传到服务端。"""
        conn = self._db._get_conn()
        conn.execute(
            "INSERT OR REPLACE INTO v2_device_keys (device_id, key_type, group_id, key_id, private_key, public_key, created_at) "
            "VALUES (?, 'group_spk_uploaded', ?, ?, ?, ?, ?)",
            (device_id, group_id, spk_id, b"", b"", int(time.time() * 1000)),
        )
        conn.commit()

    def load_latest_uploaded_group_spk_id(self, device_id: str, group_id: str) -> str | None:
        """返回本群最近一次上传成功且本地私钥仍存在的 group SPK ID。"""
        conn = self._db._get_conn()
        row = conn.execute(
            "SELECT marker.key_id FROM v2_device_keys AS marker "
            "WHERE marker.device_id=? AND marker.key_type='group_spk_uploaded' AND marker.group_id=? "
            "AND EXISTS ("
            "  SELECT 1 FROM v2_device_keys AS key "
            "  WHERE key.device_id=marker.device_id AND key.key_type='group_spk' AND key.group_id=marker.group_id AND key.key_id=marker.key_id"
            ") "
            "ORDER BY marker.created_at DESC LIMIT 1",
            (device_id, group_id),
        ).fetchone()
        return str(row[0]) if row is not None else None

    def list_recent_spk_ids(self, device_id: str, n: int) -> list[str]:
        """返回最近 N 代 SPK 的 spk_id 列表（按 created_at 降序）。

        用于销毁判定时的"最近 N 代保留窗口"——这些 SPK 即使其它条件满足也不销毁。
        """
        if n <= 0:
            return []
        conn = self._db._get_conn()
        rows = conn.execute(
            "SELECT key_id FROM v2_device_keys "
            "WHERE device_id=? AND key_type='spk' AND group_id='' ORDER BY created_at DESC LIMIT ?",
            (device_id, int(n)),
        ).fetchall()
        return [str(r[0]) for r in rows]

    def list_expired_spk_ids(self, device_id: str, max_age_seconds: float) -> list[str]:
        """返回 created_at 超过 max_age_seconds 的 SPK key_id 列表（不含当前 SPK）。"""
        import time
        cutoff = time.time() - max_age_seconds
        conn = self._db._get_conn()
        rows = conn.execute(
            "SELECT key_id FROM v2_device_keys "
            "WHERE device_id=? AND key_type='spk' AND group_id='' AND created_at < ?",
            (device_id, cutoff),
        ).fetchall()
        return [str(r[0]) for r in rows]
